### PR TITLE
feature: S3C-2791 extend copy object for object lock

### PR DESCRIPTION
**Pull request by ilkescality**
Extends COPY Object API to support object lock feature. An object should be copied from a source to a destination with additional `x-amz-object-lock-retain-until-date`, `x-amz-object-lock-mode`, `x-amz-object-lock-legal-hold` headers.

Tests cases covering the scenarios:

1. COPY object request with retention headers made to an object lock disabled destination bucket, should return `InvalidRequest` error.
`"An error occurred (InvalidRequest) when calling the CopyObject operation: Bucket is missing ObjectLockConfiguration".`
2. COPY object request without retention headers made to an object lock enabled destination bucket with default retention should copy the object with default retention of its destination bucket.
3. COPY object request without retention headers but with legal hold header made to an object lock enabled bucket with default retention should copy the object with default retention of its destination bucket.
4. COPY object request with retention headers made to an object lock enabled destination bucket with default retention should copy the object with its own retention.
5. COPY object request without retention headers made to an object lock enabled bucket without default retention should copy the object without any retention.
6. If the source object has retention configuration, the retention configuration should not be copied. (Unless either COPY object request has retention headers or destination bucket has default retention, the object should be not copied with retention configuration.)

### CODE REVIEW

File: tests/functional/aws-node-sdk/test/object/objectCopy.js [hunk 1]
```diff
@@ -1197,3 +1224,128 @@ describe('Object Copy', () => {
...
+        }).then(res => {
+            lastModified = res.LastModified;
+        }));
+
+        afterEach(() => bucketUtil.empty(sourceBucketName)
+            .then(() => bucketUtil.empty(destBucketName)));
+
+        after(() => bucketUtil.deleteMany([sourceBucketName, destBucketName]));
+
+        it('should not copy default retention info of the destination ' +
```
Comment (dora-korpar) [inline tests/functional/aws-node-sdk/test/object/objectCopy.js:1272]: I think this functionality from AWS is so weird, but good catch

**Review by dora-korpar — APPROVED** [review]

**Review by anurag4DSB — APPROVED** [review]

**Comment by ilkescality** [comment]
@bert-e approve

**Pull request merged**